_drafts/reading-legacy-code-to-harmonize-ruby-and-rails-code-loaders.md

```diff
@@ -0,0 +1,701 @@
+---
+categories:
+- metasploit-framework
+- austin.rb
+layout: post
+title: "Reading legacy code to harmonize Ruby and Rails code loaders"
+---
+
+Presented at Austin.rb in February 2015
+
+# Abstract
+
+Metasploit Framework is a a tool for developing and executing exploits code against remote target machines.  Metasploit Framework was created in 2003 in Perl and was finished converting to Ruby in 2007.  Due to its age, Metasploit Framework predates a lot of the tools and practices we take for granted like [bundler](http://bundler.io/), standard gem layout[1](http://guides.rubygems.org/make-your-own-gem/)[2](http://bundler.io/v1.7/bundle_gem.html), unit testing with [rspec](https://relishapp.com/rspec), documenting with [YARD](http://yardoc.org/), and [continuous integration](https://travis-ci.org).  In August 2012 I joined Rapid7 to work on Metasploit and my first task was to allow compatibility between Metasploit Framework's dynamic code loading of Metasploit Modules and the dynamic code loading used in Rails, `ActiveSupport::Dependencies`.  In order to make the non-Rails Metasploit Framework and Rails Metasploit Pro loading compatible I had to read and understand the Metasploit Modules loader when there were
+no tests; there were no docs; and the specifications were locked in people heads as tribal knowledge.  I also had to read `ActiveSupport::Dependencies` to understand what hidden preconditions it had on code it loaded.  Integrating this knowledge I was able to make the Metasploit Module loader compatible by dynamically assigning deterministic namespace names to the loaded Metasploit Modules.
+
+* Table of Contents
+{:toc}
+
+# Background
+
+## Code-Loading in Ruby
+
+Ruby has 3 ways to load code in the standard library:
+
+1. `load`
+2. `require`
+3. `autoload`
+
+All three search for files in the load path, which is an array of String representing directories under which relative paths can be resolved.  (If an absolute path is given the it is loaded directly without checking under directories in the load path.) The current load path can be accessed in the `$:` or the less cryptic, `$LOAD_PATH`.
```

Comment:
All three search for files in the load path, which is an array of `String` representing directories under which relative paths can be resolved.  (If an absolute path is given**, ~~the~~ then** it is loaded directly without checking under directories in the load path.) The current load path can be accessed in the `$:` **variable** or **in** the less cryptic**~~,~~ (remove comma)** `$LOAD_PATH` **variable**.